### PR TITLE
NODE-2040: migrate test runner to plain mocha and mongo-orchestration

### PR DESCRIPTION
**Pull request by mbroadst**
## Description
This change fully removes our dependency on `mongodb-test-runner` and `mongodb-version-manager`, replacing them with `mongo-orchestration` across both TravisCI and Evergreen, as well as a pure `mocha` test infrastructure.

### CODE REVIEW

File: test/runner/index.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    err => callback(err, context)
+  );
+}
+
+function filterOutTests(suite) {
+  suite.tests = suite.tests.filter(test => filters.every(f => f.filter(test)));
+  suite.suites.forEach(suite => filterOutTests(suite));
+}
+
+before(function(_done) {
```
Comment (daprahamian) [inline test/runner/index.js]: TODO: rewrite with async/await in v4

File: test/runner/metadata_ui.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
+'use strict';
```
Comment (daprahamian) [inline test/runner/metadata_ui.js]: Are there any changes here from `metamocha`?

File: test/examples/change_streams.js [hunk 3]
```diff
@@ -16,6 +16,7 @@ describe('examples(change-stream):', function() {
     client = await this.configuration.newClient().connect();
     db = client.db(this.configuration.db);
 
+    await db.createCollection('inventory');
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/examples/change_streams.js]: just confirmed that there is no error if this is called multiple times in a row

File: test/functional/mongo_client_tests.js [hunk 4]
```diff
@@ -616,11 +626,15 @@ describe('MongoClient', function() {
...
       var url = configuration
         .url()
-        .replace('rs_name=rs', '')
+        .replace('replicaSet=rs', '')
         .replace('localhost:31000', 'localhost:31000,localhost:31001');
 
       const client = configuration.newClient(url);
       client.connect(function(err) {
+        if (err == null) {
+          console.dir({ topology: client.topology }, { depth: 5 });
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/functional/mongo_client_tests.js]: nope! nice catch. fixed

File: lib/core/sdam/monitoring.js [hunk 5]
```diff
@@ -177,6 +177,12 @@ function monitorServer(server, options) {
     // emit an event indicating that our description has changed
     server.emit('descriptionReceived', new ServerDescription(server.description.address, isMaster));
 
+    const STATE_DISCONNECTED = 'disconnected';
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/core/sdam/monitoring.js]: fixed

File: test/runner/filters/mongodb_topology_filter.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    callback();
+  }
+
+  filter(test) {
+    if (!test.metadata) return true;
+    if (!test.metadata.requires) return true;
+    if (!test.metadata.requires.topology) return true;
+
+    let topologies = null;
+    if (typeof test.metadata.requires.topology === 'string') {
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/filters/mongodb_topology_filter.js]: not sure what you mean here? You mean a ternary assignment?

File: test/runner/index.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  const context = {};
+
+  eachAsync(
+    filterFiles,
+    (filterName, cb) => {
+      const FilterModule = require(path.join(__dirname, 'filters', filterName));
+      const filter = new FilterModule();
+
+      if (typeof filter !== 'object') {
+        throw new TypeError('Type of filter must be an object');
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/index.js]: fixed

File: test/runner/metadata_ui.js (hunk 2 above)
Comment (mbroadst) [reply test/runner/metadata_ui.js]: none

File: test/runner/filters/es6_generators_supported_filter.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+
+class ES6GeneratorsSupportedFilter {
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/filters/es6_generators_supported_filter.js]: removed

File: test/runner/filters/es6_promises_supported_filter.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+
+class ES6PromisesSupportedFilter {
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/filters/es6_promises_supported_filter.js]: removed

File: test/runner/config.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+  writeConcernMax() {
+    if (this.topologyType !== TopologyType.Single) {
+      return { w: 'majority', wtimeout: 30000 };
+    }
+
+    return { w: 1 };
+  }
+
+  // server37631Workaround() {
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/config.js]: deleted

File: test/runner/config.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+    // support MongoClient contructor form (url, options) for `newClient`
+    if (typeof dbOptions === 'string') {
+      return new MongoClient(
+        dbOptions,
+        this.usingUnifiedTopology()
+          ? Object.assign({ useUnifiedTopology: true }, serverOptions)
+          : serverOptions
+      );
+    }
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/config.js]: removed

File: test/runner/filters/mongodb_topology_filter.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (daprahamian) [reply test/runner/filters/mongodb_topology_filter.js]: I mean pull it into a variable:

```js
const topologyConfig = test.metadata && test.metadata.requires && test.metadata.requires.topology;
if (!topologyConfig) return true;

let topologies
if (typeof topologyConfig === 'string') {
...
```

[the remark this answers is not in the document]
Comment (mbroadst) [reply test/runner/filters/mongodb_topology_filter.js]: done

**Pull request merged**